[jit] separate QualifiedName into its own file

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19575 [jit] Use QualifiedName for classes
* #19626 [jit] make QualifiedName a value type
* **#19566 [jit] separate QualifiedName into its own file**

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

Differential Revision: [D15032205](https://our.internmc.facebook.com/intern/diff/D15032205)

suo committed:
[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Test Plan:
No behavioral changes, so existing test_jit.py tests

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Test Plan:
No behavioral changes, so existing test_jit.py tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

eellison commented on aten/src/ATen/core/qualified_name.h:
```diff
@@ -0,0 +1,92 @@
+    if (!prefix_) {
+      return !other->prefix_ && namesEqual;
+    }
+    return prefix_->equals(other->prefix_) && namesEqual;
```
Not sure if this would be optimized away, but I think it would be better to `namesEqual && prefix_->equals...` so that we don't recur down the prefixes in the case that the names aren't equal

eellison commented on test/cpp/jit/test_qualified_name.h:
```diff
@@ -0,0 +1,51 @@
+    ASSERT_ANY_THROW(QualifiedName::create("foo.bar"));
+    ASSERT_ANY_THROW(QualifiedName::createFromDotted("foo..bar"));
+    ASSERT_ANY_THROW(QualifiedName::createFromDotted(".foo.bar"));
+    ASSERT_ANY_THROW(QualifiedName::createFromDotted("foo.bar."));
```
Add empty string case ?

eellison commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -84,48 +88,33 @@ static bool isValidIdentifier(const std::string& name) {
     }
+    out << name_;
+  } else {
+    AT_ASSERT(prefix_);
```
I'm not sure i understand what is happening here

suo replied:
Just preserving what's already here. I assume this is because invalid idents can only come from Sequential, which is guaranteed to have a prefix

suo commented on aten/src/ATen/core/qualified_name.h:
```diff
@@ -0,0 +1,92 @@
+      auto atom = name.substr(startSearchFrom, pos - startSearchFrom);
+      AT_ASSERTM(
+          atom.size() > 0, "Invalid name for qualified name: '", name, "'");
+      qualifiedName = QualifiedName::create(qualifiedName, std::move(atom));
```
It's not reusing, the `create` creates a fresh pointer; I'm just reusing then name so that the loop is cleaner.

suo commented on aten/src/ATen/core/qualified_name.h:
```diff
@@ -0,0 +1,92 @@
+    size_t startSearchFrom = 0;
+    size_t pos = name.find('.', startSearchFrom);
+
+    if (pos == std::string::npos) {
```
Yeah this is redundant with the `finalAtom` logic below. I'll change.

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

eellison commented on aten/src/ATen/core/qualified_name.h:
```diff
@@ -0,0 +1,91 @@
+  QualifiedName(QualifiedNamePtr prefix, std::string name)
+      : prefix_(std::move(prefix)), name_(std::move(name)) {
+    const auto pos = name_.find('.');
+    AT_ASSERTM(
```
Is there a valid reason to have an empty string name?

suo replied:
There needs to be a null state if you want to, e.g. default construct QualifiedNames to put them in maps.

eellison reviewed (approved):
Looks good, maybe throw on empty qualified name

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head

suo committed:
Update on "[jit] separate QualifiedName into its own file"

[jit] separate QualifiedName into its own file

We want to use this to generically represent qualified names for things
(e.g. ScriptClasses). So separating out the generic stuff from the
python_print-specific stuff and sticking it in its own file.

Also added some convenience methods to go to and from fully qualified
strings.

Test Plan:
added unit tests

gh-metadata: pytorch pytorch 19566 gh/suo/29/head
